Fetch Metadata article for web.dev (#2939)
GoogleChrome/web.dev: src/site/content/en/blog/fetch-metadata/index.md

The extra information these request headers provide is quite simple, but the additional context allows you to build powerful security logic on the server-side, also referred to as a Resource Isolation Policy, with just a few lines of code.

### Implementing a Resource Isolation Policy

A Resource Isolation Policy prevents your resources from being requested by external websites. Blocking such traffic mitigates common cross-site web vulnerabilities such as CSRF, XSSI, timing attacks, and cross-origin information leaks. This policy can be enabled for all endpoints of your application and will allow all resource requests coming from your own application as well as direct navigations (via an HTTP `GET` request). Endpoints that are supposed to be loaded in a cross-site context (e.g. endpoints loaded using CORS) can be opted out of this logic.

#### Step 1: Allow requests from browsers which don't send Fetch Metadata

Since not all browsers support Fetch Metadata, you need to allow requests that don't set `Sec-Fetch-*` headers by checking for the presence of `sec-fetch-site`.

{% Aside %}
  All of the following examples are Python code.
{% endAside %}

```python

>   if not req['sec-fetch-site']:
>     return True  # Allow this request

Reviewer: ```suggestion
if not req['sec-fetch-site']:
  return True  # Allow this request
```

 ```

{% Aside 'caution' %}
Since Fetch Metadata is only supported in modern browsers, it should be used as a [defense-in-depth protection](https://static.googleusercontent.com/media/landing.google.com/en//sre/static/pdf/Building_Secure_and_Reliable_Systems.pdf#page=181) and not as your primary line of defense.
{% endAside %}

#### Step 2: Allow same-site and browser-initiated requests

Any requests that do not originate from a cross-origin context (like `evil.example`) will be allowed. In particular, these are requests that:

 -  Originate from your own application (e.g. a same-origin request where `site.example` requests `site.example/foo.json` will always be allowed).
 - Originate from your subdomains.
 -  Are explicitly caused by a user's interaction with the user agent (e.g. direct navigation or by clicking a bookmark, etc.).

```python

>   if req['sec-fetch-site'] in ('same-origin', 'same-site', 'none'):
>     return True  # Allow this request

Reviewer: ```suggestion
if req['sec-fetch-site'] in ('same-origin', 'same-site', 'none'):
  return True  # Allow this request
```

```

{% Aside 'gotchas' %}
In case your subdomains are not fully trusted, you can make the policy stricter by blocking requests from subdomains by removing the `same-site` value. 
{% endAside %} 

#### Step 3: Allow simple top-level navigation and iframing

To ensure that your site can still be linked from other sites, you have to allow simple (`HTTP GET`) top-level navigation.  

```python

>   if req['sec-fetch-mode'] == 'navigate' and req.method == 'GET'
>     # <object> and <embed> send navigation requests, which we disallow.
>     and req['sec-fetch-dest'] not in ('object', 'embed'):
>       return True  # Allow this request

Reviewer: ```suggestion
if req['sec-fetch-mode'] == 'navigate' and req.method == 'GET'
  # <object> and <embed> send navigation requests, which we disallow.
  and req['sec-fetch-dest'] not in ('object', 'embed'):
    return True  # Allow this request
```

 
From our experience of rolling out a Fetch Metadata Resource Isolation Policy at Google, most applications are by default compatible with such a policy and rarely require exempting endpoints to allow cross-site traffic.

### Enforcing a Resource Isolation Policy
After you've checked that your policy doesn't impact legitimate production traffic, you're ready to enforce restrictions, guaranteeing that other sites will not be able to request your resources and protecting your users from cross-site attacks.

{% Aside 'caution' %}
Make sure that you reject invalid requests before running authentication checks or any other processing of the request to prevent revealing sensitive timing information. 
{% endAside %}

## Further reading

- [W3C Fetch Metadata Request Headers specification](https://www.w3.org/TR/fetch-metadata/)
- [Fetch Metadata Playground](https://secmetadata.appspot.com/)

> - [Google I/O - Securing Web Apps with Modern Platform Features](https://webappsec.dev/assets/pub/Google_IO-Securing_Web_Apps_with_Modern_Platform_Features.pdf)

Reviewer: ```suggestion
- [Google I/O - Securing Web Apps with Modern Platform Features](https://webappsec.dev/assets/pub/Google_IO-Securing_Web_Apps_with_Modern_Platform_Features.pdf) (Slides)
```

 - `none`, if the request was explicitly caused by a user's interaction with the user agent (e.g. clicking on a bookmark)
 - `cross-site`, if the request was sent by another website (e.g. `evil.example`)

### `Sec-Fetch-Mode`

`Sec-Fetch-Mode` indicates the [mode](https://developer.mozilla.org/docs/Web/API/Request/mode) of the request. This roughly corresponds to the type of the request and allows you to distinguish resource loads from navigation requests. For example, a destination of `navigate` indicates a top-level navigation request while `no-cors` indicates resource requests like loading an image.

### `Sec-Fetch-Dest`

`Sec-Fetch-Dest` exposes a request's [destination](https://developer.mozilla.org/docs/Web/API/Request/destination) (e.g. if a `script` or an `img` tag caused a resource to be requested by the browser).

## How to use Fetch Metadata to protect against cross-origin attacks

The extra information these request headers provide is quite simple, but the additional context allows you to build powerful security logic on the server-side, also referred to as a Resource Isolation Policy, with just a few lines of code.

### Implementing a Resource Isolation Policy

> A Resource Isolation Policy prevents your resources from being requested by external websites. Blocking such traffic mitigates common cross-site web vulnerabilities such as CSRF, XSSI, timing attacks, and cross-origin information leaks. This policy can be enabled for all endpoints of your application and will allow all resource requests coming from your own application as well as direct navigations (via an HTTP `GET` request). Endpoints that are supposed to be loaded in a cross-site context (e.g. endpoints loaded using CORS) can be opted out of this logic.

Reviewer: I can't find any grammar articles about "noun phrases being spliced together" and I think the meaning of the sentence is clear enough so let's just leave as is

      return True  # Allow this request
```
{% Aside 'gotchas' %}
The logic above protects your application's endpoints from being used as resources by other websites, but will permit top-level navigation and embedding (e.g. loading in an <iframe>). To further improve security, you can use Fetch Metadata headers to restrict cross-site navigations to only an allowed set of pages.
{% endAside %} 

#### Step 4: Opt out endpoints that are meant to serve cross-site traffic (Optional) 

In some cases, your application might provide resources which are meant to be loaded cross-site. These resources need to be exempted on a per-path or per-endpoint basis. Examples of such endpoints are:

 - Endpoints meant to be accessed cross-origin - If your application is serving endpoints that are `CORS` enabled, you need to explicitly opt them out from resource isolation to ensure that cross-site requests to these endpoints are still possible.

>  - Public resources (e.g. images, styles, etc) - Any public and unauthenticated resources that should be loadable cross-origin from other sites can be exempted as well.

Reviewer 2: ```suggestion
 - Public resources (e.g. images, styles, etc.) - Any public and unauthenticated resources that should be loadable cross-origin from other sites can be exempted as well.
```

```python

>   if req.path in ('/my_CORS_endpoint', '/favicon.png'):
>     return True

Reviewer 2: ```suggestion
if req.path in ('/my_CORS_endpoint', '/favicon.png'):
  return True
```

Reviewer 2: Can you please remove the indentation from all code examples? It's causing the page to render with the wrong left-hand spacing.

{% endAside %}

#### Step 5: Reject all other requests that are cross-site and not navigational

Any other **cross-site** request will be rejected by this Resource Isolation Policy and thus protect your application from common cross-site attacks.

{% Aside 'gotchas' %}
By default, requests violating your policy should be rejected with an `HTTP 403` response. But, depending on your use case, you can also consider other actions, such as:
 - **Only logging violations**. This is especially useful when testing the compatibility of the policy and finding endpoints that might need to be opted out.
 - **Modifying the request**. In certain scenarios, consider performing other actions like redirecting to your landing page and dropping authentication credentials (e.g. cookies). However, be aware that this could weaken the protections of a Fetch Metadata-based policy.
{% endAside %}

**Example:** The following code demonstrates a complete implementation of a robust Resource Isolation Policy on the server or as a middleware to deny potentially malicious cross-site resource requests, while allowing simple navigational requests:

```python

> # Reject cross-origin requests to protect from CSRF, XSSI, & other bugs

Reviewer 2: ```suggestion
# Reject cross-origin requests to protect from CSRF, XSSI, and other bugs
```

      return True

  # [OPTIONAL] Exempt paths/endpoints meant to be served cross-origin.
  if req.path in ('/my_CORS_endpoint', '/favicon.png'):
    return True

  # Reject all other requests that are cross-site and not navigational
  return False
```

### Deploying a Resource Isolation Policy

1. Install a module like the code snippet from above to log and monitor how your site behaves and make sure the restrictions don't affect any legitimate traffic.
1. Fix potential violations by exempting legitimate cross-origin endpoints.
1. Enforce the policy by dropping non-compliant requests.

### Identifying and fixing policy violations

> It's recommended to test your policy in a side-effect free way by first enabling it in reporting mode in your server-side code. Alternatively, you can implement this logic in middleware, or in a reverse proxy which logs any violations that your policy might produce when applied to production traffic.

Reviewer 2: ```suggestion
It's recommended that you test your policy in a side-effect free way by first enabling it in reporting mode in your server-side code. Alternatively, you can implement this logic in middleware, or in a reverse proxy which logs any violations that your policy might produce when applied to production traffic.
```

 
From our experience of rolling out a Fetch Metadata Resource Isolation Policy at Google, most applications are by default compatible with such a policy and rarely require exempting endpoints to allow cross-site traffic.

### Enforcing a Resource Isolation Policy

> After you've checked that your policy doesn't impact legitimate production traffic, you're ready to start enforcing the restrictions, guaranteeing that other sites will not be able to request your resources and protecting your users from cross-site attacks.

Reviewer 2: ```suggestion
After you've checked that your policy doesn't impact legitimate production traffic, you're ready to enforce restrictions, guaranteeing that other sites will not be able to request your resources and protecting your users from cross-site attacks.
```
